README
NAME
    Setup::File::Symlink - Setup symlink (existence, target)

VERSION
    This document describes version 0.26 of Setup::File::Symlink (from Perl
    distribution Setup-File-Symlink), released on 2014-05-17.

FUNCTIONS
  ln_s(%args) -> [status, msg, result, meta]
    Create symlink.

    Fixed state: "symlink" exists and points to "target".

    Fixable state: "symlink" doesn't exist.

    This function is idempotent (repeated invocations with same arguments
    has the same effect as single invocation). This function supports
    transactions.

    Arguments ('*' denotes required arguments):

    *   symlink => *str*

        Path to symlink.

    *   target => *str*

        Path to target.

    Special arguments:

    *   -tx_action => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_action_id => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_recovery => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_rollback => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_v => *str*

        For more information on transaction, see Rinci::Transaction.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

  rmsym(%args) -> [status, msg, result, meta]
    Delete symlink.

    Will not delete non-symlinks.

    Fixed state: "path" doesn't exist.

    Fixable state: "path" exists, is a symlink, (and if "target" is defined,
    points to "target").

    This function is idempotent (repeated invocations with same arguments
    has the same effect as single invocation). This function supports
    transactions.

    Arguments ('*' denotes required arguments):

    *   path => *str*

    *   target => *str*

        Only delete if existing symlink has this target.

    Special arguments:

    *   -tx_action => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_action_id => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_recovery => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_rollback => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_v => *str*

        For more information on transaction, see Rinci::Transaction.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

  setup_symlink(%args) -> [status, msg, result, meta]
    Setup symlink (existence, target).

    When "should_exist="1> (the default): On do, will create symlink which
    points to specified target. If symlink already exists but points to
    another target, it will be replaced with the correct symlink if
    "replace_symlink" option is true. If a file/dir already exists and
    "replace_file"/"replace_dir" option is true, it will be moved (trashed)
    first before the symlink is created. On undo, will delete symlink if it
    was created by this function, and restore the original symlink/file/dir
    if it was replaced during do.

    When "should_exist="0>: On do, will remove symlink if it exists (and
    "replace_symlink" is true). If "replace_file"/"replace_dir" is true,
    will also remove file/dir. On undo, will restore deleted
    symlink/file/dir.

    This function is idempotent (repeated invocations with same arguments
    has the same effect as single invocation). This function supports
    transactions.

    Arguments ('*' denotes required arguments):

    *   create => *bool* (default: 1)

        Create if symlink doesn't exist.

        If set to false, then setup will fail (412) if this condition is
        encountered.

    *   replace_dir => *bool* (default: 0)

        Replace if there is existing dir.

        If set to false, then setup will fail (412) if this condition is
        encountered.

    *   replace_file => *bool* (default: 0)

        Replace if there is existing non-symlink file.

        If set to false, then setup will fail (412) if this condition is
        encountered.

    *   replace_symlink => *bool* (default: 1)

        Replace previous symlink if it already exists but doesn't point to
        the wanted target.

        If set to false, then setup will fail (412) if this condition is
        encountered.

    *   should_exist => *bool* (default: 1)

        Whether symlink should exist.

    *   symlink* => *str*

        Path to symlink.

    *   target => *str*

        Target path of symlink.

        Required, unless "should_exist =" 0>.

    Special arguments:

    *   -tx_action => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_action_id => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_recovery => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_rollback => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_v => *str*

        For more information on transaction, see Rinci::Transaction.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

SEE ALSO
    Setup

    Setup::File

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/Setup-File-Symlink>.

SOURCE
    Source repository is at
    <https://github.com/sharyanto/perl-Setup-File-Symlink>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Setup-File-Symlink>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Setup-File-Symlink

0.26    2014-05-17 (SHARYANTO)

        - No functional changes.

        - Replace File::Slurp with File::Slurp::Tiny.


0.25    2013-08-05 (SHARYANTO)

        - No functional changes. Tweak error messages to give more
          details.


0.24    2012-08-30 (SHARYANTO)

        [ENHANCEMENTS]

        - Add argument 'should_exist'.


0.23    2012-08-28 (SHARYANTO)

        - Do not require symlink path to be absolute.


0.22    2012-08-24 (SHARYANTO)

        [ENHANCEMENTS]

        - Now uses Rinci::Transaction protocol v=2, no longer uses
          Perinci::Sub::Gen::Undoable.

        - Add functions: ln_s(), rmsym().


0.21    2012-07-04 (SHARYANTO)

        - No functional changes. Adjust to Perinci::Sub::Gen::Undoable
          0.09.


0.20    2012-06-23 (SHARYANTO)

        - No functional changes. Adjust to Perinci::Sub::Gen::Undoable
          0.08.


0.19    2012-06-23 (SHARYANTO)

        - No functional changes. Adjust to Perinci::Sub::Gen::Undoable
          0.07.


0.18    2012-06-22 (SHARYANTO)

        - No functional changes. Adjust to Perinci::Sub::Gen::Undoable
          0.06.


0.17    2012-06-22 (SHARYANTO)

        - No functional changes.

        - rinci: Fix argument position.


0.16    2012-06-06 (SHARYANTO)

        - No functional changes. Now uses Perinci::Sub::Gen::Undoable.


0.15    2012-03-28 (SHARYANTO)

        - No functional changes. Require newer Test::Setup.


0.14    2012-03-28 (SHARYANTO)

        - No functional changes. Replace SubSpec PodWeaver plugin with
          Perinci.  Use Test::Setup.


0.13    2012-03-02 (SHARYANTO)

        - No functional changes. dzil tweak to avoid line number change.


0.12    2012-03-02 (SHARYANTO)

        - No functional changes. Replace references to Sub::Spec now
          to Rinci.


0.11    2011-05-18 (SHARYANTO)

        - No functional changes. Tweak log messages.


0.10    2011-05-18 (SHARYANTO)

        - No functional changes. Changed logging levels.


0.09    2011-04-28 (SHARYANTO)

        [INCOMPATIBLE CHANGES]

        - Update undo protocol to Sub::Spec 0.13, there's no longer redo_data,
          only undo_data.

        [BUG FIXES]

        - Small fixes.


0.08    2011-04-27 (SHARYANTO)

        [BUG FIXES]

        - Return 500 on rollback.

        - A couple of small fixes.


0.07    2011-04-27 (SHARYANTO)

        [ENHANCEMENTS]

        - Increase robustness, each step now also checks state and skips if
          desired state already reached. This means that undo can be reapplied
          without erroring. We cannot guarantee state changes between invocation
          of do and undo anyway, so we try our best at each step.


0.06    2011-04-27 (SHARYANTO)

        [ENHANCEMENTS]

        - (Internal) refactor/reorganization.

        - Implement replace_file & replace_dir.


0.05    2011-04-21 (SHARYANTO)

        - No functional change.

        [INCOMPATIBLE CHANGE]

        - Rename module from Setup::Symlink to Setup::File::Symlink.


0.04    2011-04-21 (SHARYANTO)

        - No functional change.

        [ETC]

        - Change Abstract to be more consistent with other Setup::*
          modules.


0.03    2011-04-18 (SHARYANTO)

        [ETC]

        - Some sub spec tweaks.

        - Change terminology: undo/redo info -> undo/redo data.


0.02    2011-04-13 (SHARYANTO)

        [ETC]

        - Minor documentation & sub spec patches.

        - Build fixes: exclude backup files, add missing dependency
          (File::chdir, Log::Any) [cpants]


0.01    2011-04-08 (SHARYANTO)

        - First release.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Setup symlink (existence, target)",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Setup-File-Symlink",
  "dist_version" => "0.25",
  "license" => "perl",
  "module_name" => "Setup::File::Symlink",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Capture::Tiny" => 0,
    "File::Trash::Undoable" => "0.06",
    "Log::Any" => 0,
    "Rinci" => "v1.1.29",
    "Setup" => "v1.0.4",
    "UUID::Random" => 0,
    "perl" => "5.010001"
  },
  "script_files" => [],
  "test_requires" => {
    "Capture::Tiny" => 0,
    "File::Slurp" => 0,
    "File::chdir" => 0,
    "Test::More" => "0.96",
    "Test::Perinci::Tx::Manager" => "0.40",
    "blib" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Setup symlink (existence, target)",
  "AUTHOR" => "Steven Haryanto <stevenharyanto\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Setup-File-Symlink",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Setup::File::Symlink",
  "PREREQ_PM" => {
    "File::Trash::Undoable" => "0.06",
    "Log::Any" => 0,
    "Rinci" => "v1.1.29",
    "Setup" => "v1.0.4",
    "UUID::Random" => 0
  },
  "TEST_REQUIRES" => {
    "File::Slurp::Tiny" => 0,
    "File::Spec" => 0,
    "File::chdir" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96",
    "Test::Perinci::Tx::Manager" => "0.40"
  },
  "VERSION" => "0.26",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "File::Slurp::Tiny" => 0,
  "File::Spec" => 0,
  "File::Trash::Undoable" => "0.06",
  "File::chdir" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Log::Any" => 0,
  "Rinci" => "v1.1.29",
  "Setup" => "v1.0.4",
  "Test::More" => "0.96",
  "Test::Perinci::Tx::Manager" => "0.40",
  "UUID::Random" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Setup/File/Symlink.pm
package Setup::File::Symlink;

use 5.010001;
use strict;
use warnings;
use Log::Any '$log';

use File::Trash::Undoable;

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(setup_symlink);

our $VERSION = '0.26'; # VERSION

our %SPEC;

$SPEC{rmsym} = {
    v           => 1.1,
    summary     => 'Delete symlink',
    description => <<'_',

Will not delete non-symlinks.

Fixed state: `path` doesn't exist.

Fixable state: `path` exists, is a symlink, (and if `target` is defined, points
to `target`).

_
    args        => {
        path => {
            schema => 'str*',
        },
        target => {
            summary => 'Only delete if existing symlink has this target',
            schema => 'str*',
        },
    },
    features => {
        tx => {v=>2},
        idempotent => 1,
    },
};
sub rmsym {
    my %args = @_;

    # TMP, schema
    my $tx_action = $args{-tx_action} // '';
    my $dry_run   = $args{-dry_run};
    my $path      = $args{path};
    defined($path) or return [400, "Please specify path"];
    my $target    = $args{target};

    my $is_sym    = (-l $path);
    my $exists    = $is_sym || (-e _);
    my $curtarget; $curtarget = readlink($path) if $is_sym;

    my @undo;

    if ($tx_action eq 'check_state') {
        return [412, "$path is not a symlink"] if $exists && !$is_sym;
        return [412, "Target of symlink $path does not match ($curtarget)"]
            if $is_sym && defined($target) && $curtarget ne $target;
        if ($exists) {
            unshift @undo, ['ln_s', {
                symlink => $path,
                target  => $target // $curtarget,
            }];
        }
        if (@undo) {
            $log->info("(DRY) Deleting symlink $path ...") if $dry_run;
            return [200, "Symlink $path should be removed", undef,
                    {undo_actions=>\@undo}];
        } else {
            return [304, "Symlink $path already does not exist"];
        }
    } elsif ($tx_action eq 'fix_state') {
        $log->info("Deleting symlink $path ...");
        if (unlink $path) {
            return [200, "OK"];
        } else {
            return [500, "Can't remove symlink $path: $!"];
        }
    }
    [400, "Invalid -tx_action"];
}

$SPEC{ln_s} = {
    v           => 1.1,
    summary     => 'Create symlink',
    description => <<'_',

Fixed state: `symlink` exists and points to `target`.

Fixable state: `symlink` doesn't exist.

_
    args        => {
        symlink => {
            summary => 'Path to symlink',
            schema => 'str*',
        },
        target => {
            summary => 'Path to target',
            schema => 'str*',
        },
    },
    features => {
        tx => {v=>2},
        idempotent => 1,
    },
};
sub ln_s {
    my %args = @_;

    # TMP, schema
    my $tx_action = $args{-tx_action} // '';
    my $dry_run   = $args{-dry_run};
    my $symlink   = $args{symlink};
    defined($symlink) or return [400, "Please specify symlink"];
    my $target    = $args{target};
    defined($target) or return [400, "Please specify target"];

    my $is_sym    = (-l $symlink);
    my $exists    = $is_sym || (-e _);
    my $curtarget; $curtarget = readlink($symlink) if $is_sym;
    my @undo;

    if ($tx_action eq 'check_state') {
        return [412, "Path $symlink already exists"] if $exists && !$is_sym;
        return [412, "Symlink $symlink points to another target"] if $is_sym &&
            $curtarget ne $target;
        if (!$exists) {
            unshift @undo, ['rmsym', {path => $symlink}];
        }
        if (@undo) {
        $log->info("(DRY) Creating symlink $symlink -> $target ...")
            if $dry_run;
        return [200, "Symlink $symlink needs to be created", undef,
                {undo_actions=>\@undo}];
        } else {
            return [304, "Symlink $symlink already exists"];
        }
    } elsif ($tx_action eq 'fix_state') {
        $log->info("Creating symlink $symlink -> $target ...");
        if (symlink $target, $symlink) {
            return [200, "Fixed"];
        } else {
            return [500, "Can't symlink $symlink -> $target: $!"];
        }
    }
    [400, "Invalid -tx_action"];
}

$SPEC{setup_symlink} = {
    v           => 1.1,
    summary     => "Setup symlink (existence, target)",
    description => <<'_',

When `should_exist=>1` (the default): On do, will create symlink which points to
specified target. If symlink already exists but points to another target, it
will be replaced with the correct symlink if `replace_symlink` option is true.
If a file/dir already exists and `replace_file`/`replace_dir` option is true, it
will be moved (trashed) first before the symlink is created. On undo, will
delete symlink if it was created by this function, and restore the original
symlink/file/dir if it was replaced during do.

When `should_exist=>0`: On do, will remove symlink if it exists (and
`replace_symlink` is true). If `replace_file`/`replace_dir` is true, will also
remove file/dir. On undo, will restore deleted symlink/file/dir.

_
    args        => {
        should_exist => {
            summary => "Whether symlink should exist",
            schema => ['bool' => {default => 1}],
        },
        symlink => {
            summary => 'Path to symlink',
            schema => ['str*' => {match => qr!^/!}],
            req => 1,
            pos => 0,
        },
        target => {
            summary => 'Target path of symlink',
            schema => 'str*',
            req => 0, # XXX only when should_exist=1
            description => <<'_',

Required, unless `should_exist => 0`.

_
        },
        create => {
            summary => "Create if symlink doesn't exist",
            schema => [bool => {default=>1}],
            description => <<'_',

If set to false, then setup will fail (412) if this condition is encountered.

_
        },
        replace_symlink => {
            summary => "Replace previous symlink if it already exists ".
                "but doesn't point to the wanted target",
            schema => ['bool' => {default => 1}],
            description => <<'_',

If set to false, then setup will fail (412) if this condition is encountered.

_
        },
        replace_file => {
            summary => "Replace if there is existing non-symlink file",
            schema => ['bool' => {default => 0}],
            description => <<'_',

If set to false, then setup will fail (412) if this condition is encountered.

_
        },
        replace_dir => {
            summary => "Replace if there is existing dir",
            schema => ['bool' => {default => 0}],
            description => <<'_',

If set to false, then setup will fail (412) if this condition is encountered.

_
        },
    },
    features    => {
        tx => {v=>2},
        idempotent => 1,
    },
};
sub setup_symlink {
    require UUID::Random;

    my %args = @_;

    # TMP, schema
    my $tx_action    = $args{-tx_action} // '';
    my $dry_run      = $args{-dry_run};
    my $should_exist = $args{should_exist} // 1;
    my $symlink      = $args{symlink} or return [400, "Please specify symlink"];
    my $target       = $args{target};
    if ($should_exist) {
        defined($target) or return [400, "Please specify target"];
    }
    my $create          = $args{create}       // 1;
    my $replace_file    = $args{replace_file} // 0;
    my $replace_dir     = $args{replace_dir}  // 0;
    my $replace_symlink = $args{replace_symlink} // 1;

    my $is_sym     = (-l $symlink); # -l performs lstat()
    my $exists     = (-e _);    # now we can use -e
    my $is_dir     = (-d _);
    my $cur_target = $is_sym ? readlink($symlink) : "";

    my $taid       = $args{-tx_action_id} // UUID::Random::generate();
    my $suffix     = substr($taid,0,8);

    my (@do, @undo);

    if ($should_exist) {
        if ($exists && !$is_sym) {
            if ($is_dir && !$replace_dir) {
                return [412, "Must replace dir $symlink with symlink ".
                            "but instructed not to"];
            } elsif (!$is_dir && !$replace_file) {
                return [412, "Must replace file $symlink with symlink ".
                            "but instructed not to"];
            }
            $log->info("(DRY) Replacing file/dir $symlink with symlink ...")
                if $dry_run;
            push @do, (
                ["File::Trash::Undoable::trash",
                 {path=>$symlink, suffix=>$suffix}],
                ["ln_s", {symlink=>$symlink, target=>$target}],
            );
            unshift @undo, (
                ["rmsym", {path=>$symlink, target=>$target}],
                ["File::Trash::Undoable::untrash",
                 {path=>$symlink, suffix=>$suffix}],
            );
        } elsif ($is_sym && $cur_target ne $target) {
            if (!$replace_symlink) {
                return [412, "Must replace symlink $symlink ".
                            "but instructed not to"];
            }
            $log->info("(DRY) Replacing symlink $symlink ...") if $dry_run;
            push @do, (
                [rmsym => {path=>$symlink}],
                [ln_s  => {symlink=>$symlink, target=>$target}],
            );
            unshift @undo, (
                ["rmsym", {path=>$symlink, target=>$target}],
                ["ln_s", {symlink=>$symlink, target=>$cur_target}],
            );
        } elsif (!$exists) {
            if (!$create) {
                return [412, "Must create symlink $symlink ".
                            "but instructed not to"];
            }
            $log->info("(DRY) Creating symlink $symlink ...") if $dry_run;
            push @do, (
                ["ln_s", {symlink=>$symlink, target=>$target}],
            );
            unshift @undo, (
                ["rmsym", {path=>$symlink}],
            );
        }
    } elsif ($exists) {
        return [412, "Must delete symlink $symlink but instructed not to"]
            if $is_sym && !$replace_symlink;
        return [412, "Must delete dir $symlink but instructed not to"]
            if $is_dir && !$replace_dir;
        return [412, "Must delete file $symlink but instructed not to"]
            if !$is_sym && !$is_dir && !$replace_file;
        $log->info("(DRY) Removing symlink $symlink ...") if $dry_run;
        push    @do  , ["File::Trash::Undoable::trash",
                        {path=>$symlink, suffix=>$suffix}];
        unshift @undo, ["File::Trash::Undoable::untrash",
                        {path=>$symlink, suffix=>$suffix}];
    }

    if (@do) {
        return [200, "", undef, {do_actions=>\@do, undo_actions=>\@undo}];
    } else {
        return [304, "Already fixed"];
    }
}

1;
# ABSTRACT: Setup symlink (existence, target)

__END__

=pod

=encoding UTF-8

=head1 NAME

Setup::File::Symlink - Setup symlink (existence, target)

=head1 VERSION

This document describes version 0.26 of Setup::File::Symlink (from Perl distribution Setup-File-Symlink), released on 2014-05-17.

=head1 FUNCTIONS


=head2 ln_s(%args) -> [status, msg, result, meta]

Create symlink.

Fixed state: C<symlink> exists and points to C<target>.

Fixable state: C<symlink> doesn't exist.

This function is idempotent (repeated invocations with same arguments has the same effect as single invocation). This function supports transactions.


Arguments ('*' denotes required arguments):

=over 4

=item * B<symlink> => I<str>

Path to symlink.

=item * B<target> => I<str>

Path to target.

=back

Special arguments:

=over 4

=item * B<-tx_action> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_action_id> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_recovery> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_rollback> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_v> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=back

Return value:

Returns an enveloped result (an array).

First element (status) is an integer containing HTTP status code
(200 means OK, 4xx caller error, 5xx function error). Second element
(msg) is a string containing error message, or 'OK' if status is
200. Third element (result) is optional, the actual result. Fourth
element (meta) is called result metadata and is optional, a hash
that contains extra information.


=head2 rmsym(%args) -> [status, msg, result, meta]

Delete symlink.

Will not delete non-symlinks.

Fixed state: C<path> doesn't exist.

Fixable state: C<path> exists, is a symlink, (and if C<target> is defined, points
to C<target>).

This function is idempotent (repeated invocations with same arguments has the same effect as single invocation). This function supports transactions.


Arguments ('*' denotes required arguments):

=over 4

=item * B<path> => I<str>

=item * B<target> => I<str>

Only delete if existing symlink has this target.

=back

Special arguments:

=over 4

=item * B<-tx_action> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_action_id> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_recovery> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_rollback> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_v> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=back

Return value:

Returns an enveloped result (an array).

First element (status) is an integer containing HTTP status code
(200 means OK, 4xx caller error, 5xx function error). Second element
(msg) is a string containing error message, or 'OK' if status is
200. Third element (result) is optional, the actual result. Fourth
element (meta) is called result metadata and is optional, a hash
that contains extra information.


=head2 setup_symlink(%args) -> [status, msg, result, meta]

Setup symlink (existence, target).

When C<should_exist=>1> (the default): On do, will create symlink which points to
specified target. If symlink already exists but points to another target, it
will be replaced with the correct symlink if C<replace_symlink> option is true.
If a file/dir already exists and C<replace_file>/C<replace_dir> option is true, it
will be moved (trashed) first before the symlink is created. On undo, will
delete symlink if it was created by this function, and restore the original
symlink/file/dir if it was replaced during do.

When C<should_exist=>0>: On do, will remove symlink if it exists (and
C<replace_symlink> is true). If C<replace_file>/C<replace_dir> is true, will also
remove file/dir. On undo, will restore deleted symlink/file/dir.

This function is idempotent (repeated invocations with same arguments has the same effect as single invocation). This function supports transactions.


Arguments ('*' denotes required arguments):

=over 4

=item * B<create> => I<bool> (default: 1)

Create if symlink doesn't exist.

If set to false, then setup will fail (412) if this condition is encountered.

=item * B<replace_dir> => I<bool> (default: 0)

Replace if there is existing dir.

If set to false, then setup will fail (412) if this condition is encountered.

=item * B<replace_file> => I<bool> (default: 0)

Replace if there is existing non-symlink file.

If set to false, then setup will fail (412) if this condition is encountered.

=item * B<replace_symlink> => I<bool> (default: 1)

Replace previous symlink if it already exists but doesn't point to the wanted target.

If set to false, then setup will fail (412) if this condition is encountered.

=item * B<should_exist> => I<bool> (default: 1)

Whether symlink should exist.

=item * B<symlink>* => I<str>

Path to symlink.

=item * B<target> => I<str>

Target path of symlink.

Required, unless C<should_exist => 0>.

=back

Special arguments:

=over 4

=item * B<-tx_action> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_action_id> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_recovery> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_rollback> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_v> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=back

Return value:

Returns an enveloped result (an array).

First element (status) is an integer containing HTTP status code
(200 means OK, 4xx caller error, 5xx function error). Second element
(msg) is a string containing error message, or 'OK' if status is
200. Third element (result) is optional, the actual result. Fourth
element (meta) is called result metadata and is optional, a hash
that contains extra information.

=head1 SEE ALSO

L<Setup>

L<Setup::File>

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/Setup-File-Symlink>.

=head1 SOURCE

Source repository is at L<https://github.com/sharyanto/perl-Setup-File-Symlink>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=Setup-File-Symlink>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
